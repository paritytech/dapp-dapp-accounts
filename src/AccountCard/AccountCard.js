// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { bonds } from 'oo7-parity';

// TODO Wait for PR#18 on parity-reactive-ui for dependency
import EtherBalance from '../vendor/EtherBalance';

import IdentityIcon from '@parity/ui/lib/IdentityIcon';
import { IdentityName } from '@parity/ui/lib/IdentityName/identityName';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';
import Image from 'semantic-ui-react/dist/commonjs/elements/Image';
import Label from 'semantic-ui-react/dist/commonjs/elements/Label';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';

import styles from './AccountCard.css';

class AccountCard extends PureComponent {
  render() {
    const {
      account,
      address,
      isDefault,
      isVisible,
      onMakeDefault,
      onToggleVisible
    } = this.props;

    return (
      <Card
        className={[styles.accountCard, !isVisible ? styles.hidden : ''].join(
          ' '
        )}
        onClick={isVisible ? null : () => onToggleVisible(address)}
      >
        <Card.Content>
          <Image
            as={IdentityIcon}
            floated="right"
            size="mini"
            address={address}
          />
          <Card.Header>
            <IdentityName address={address} name={account.name} unknown />
          </Card.Header>
          <Card.Meta>{address}</Card.Meta>
          <Card.Description>
            {account.meta.description && <p>{account.meta.description}</p>}
            <p>
              <FormattedMessage
                id="dapp.accounts.balance"
                defaultMessage="Balance"
              />: <EtherBalance balance={bonds.balance(address)} />
            </p>
          </Card.Description>
        </Card.Content>
        {!isDefault && (
          <Card.Content extra textAlign="center">
            {isVisible ? (
              <Button.Group widths="2">
                <Button basic primary onClick={() => onMakeDefault(address)}>
                  Make Default
                </Button>
                <Button
                  basic
                  color="grey"
                  onClick={() => onToggleVisible(address)}
                >
                  Hide
                </Button>
              </Button.Group>
            ) : (
              <Label>
                <FormattedMessage
                  id="dapp.accounts.hidden"
                  defaultMessage="Hidden"
                />
              </Label>
            )}
          </Card.Content>
        )}
      </Card>
    );
  }
}

export default AccountCard;

AccountCard.propTypes = {
  account: PropTypes.object.isRequired,
  address: PropTypes.string.isRequired,
  isDefault: PropTypes.bool,
  isVisible: PropTypes.bool.isRequired,
  onMakeDefault: PropTypes.func,
  onToggleVisible: PropTypes.func
};
