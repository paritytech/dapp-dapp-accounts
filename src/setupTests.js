// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { IntlProvider, intlShape } from 'react-intl';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Configure enzyme with React 16
configure({ adapter: new Adapter() });

/**
 * Components using the react-intl module require access to the intl context.
 * This is not available when mounting single components in Enzyme.
 * These helper functions aim to address that and wrap a valid,
 * English-locale intl context around them.
 */

// Create the IntlProvider to retrieve context for wrapping around.
const intlProvider = new IntlProvider({ locale: 'en' }, {});
const { intl } = intlProvider.getChildContext();

/**
 * When using React-Intl `injectIntl` on components, props.intl is required.
 */
const nodeWithIntlProp = node => {
  return React.cloneElement(node, { intl });
};

export const shallowWithIntl = (node, { context } = {}) => {
  return shallow(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl })
  });
};

export const mountWithIntl = (node, { context, childContextTypes } = {}) => {
  return mount(nodeWithIntlProp(node), {
    context: Object.assign({}, context, { intl }),
    childContextTypes: Object.assign({}, { intl: intlShape }, childContextTypes)
  });
};
