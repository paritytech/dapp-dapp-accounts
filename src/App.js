// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import { observer } from 'mobx-react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';

import Page from '@parity/ui/lib/Page';
import Header from 'semantic-ui-react/dist/commonjs/elements/Header';
import Card from 'semantic-ui-react/dist/commonjs/views/Card';

import styles from './App.css';
import AccountCard from './AccountCard';

class App extends Component {
  handleMakeDefault = address => {
    const { store } = this.props;
    store.saveDefault(address);
  };

  handleToggleVisible = address => {
    const { store } = this.props;
    if (store.visible.includes(address)) {
      store.hideAccount(address);
    } else {
      store.showAccount(address);
    }
  };

  render() {
    const { store } = this.props;

    if (!Object.keys(store.accounts) || !store.default) return null;

    return (
      <Page
        title={
          <FormattedMessage
            id="dapps.accounts.title"
            defaultMessage="Accounts Visible to Dapps"
          />
        }
      >
        <div className={styles.section}>
          <Header as="h4">Default Account</Header>
          <span>
            This is your default account to be used in the Dapps. It is also the
            account shown on the top bar.
          </span>
          <Card.Group className={styles.cardGroup}>
            <AccountCard
              isDefault
              isVisible
              account={store.accounts[store.default]}
              address={store.default}
            />
          </Card.Group>
        </div>
        <div className={styles.section}>
          <Header as="h4">Other Accounts</Header>
          <span>
            These are the other accounts that can be accessed by Dapps. Toggle
            them to make them visible/hidden for Dapps.
          </span>
          <Card.Group className={styles.cardGroup}>
            {Object.keys(store.accounts)
              .filter(address => address !== store.default)
              .map(address => (
                <AccountCard
                  key={address}
                  address={address}
                  account={store.accounts[address]}
                  isVisible={store.visible.includes(address)}
                  onMakeDefault={this.handleMakeDefault}
                  onToggleVisible={this.handleToggleVisible}
                />
              ))}
          </Card.Group>
        </div>
      </Page>
    );
  }
}

export default observer(App);

App.propTypes = {
  store: PropTypes.object.isRequired
};
