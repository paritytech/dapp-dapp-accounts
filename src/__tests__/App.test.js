// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallowToJson } from 'enzyme-to-json';

import { shallowWithIntl } from '../setupTests';
import App from '../App';
import AccountCard from '../AccountCard';

const props = {
  // Mock store
  store: {
    accounts: {
      foo: {
        name: 'FooName',
        meta: {}
      },
      bar: {
        name: 'BarName',
        meta: {}
      },
      baz: {
        name: 'BazName',
        meta: {}
      }
    },
    visible: ['foo', 'bar'],
    default: 'foo'
  }
};

test('should render correctly', () => {
  const component = shallowWithIntl(<App {...props} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly when no account', () => {
  const component = shallowWithIntl(<App store={{ accounts: {} }} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should call store.showAccount when we toggle visibility on a hidden app', () => {
  const showAccount = jest.fn();
  const component = shallowWithIntl(
    <App store={{ ...props.store, showAccount }} />
  );

  component
    .find(AccountCard)
    .last() // baz is hidden
    .props()
    .onToggleVisible();

  expect(showAccount).toHaveBeenCalled();
});

test('should call store.hideAccount when we toggle visibility on a visible app', () => {
  const hideAccount = jest.fn();
  const component = shallowWithIntl(
    <App store={{ ...props.store, hideAccount }} />
  );

  const componentProps = component
    .find(AccountCard)
    .at(1) // bar is shown
    .props();

  componentProps.onToggleVisible(componentProps.address);

  expect(hideAccount).toHaveBeenCalled();
});

test('should call store.saveDefault when we make an account default', () => {
  const saveDefault = jest.fn();
  const component = shallowWithIntl(
    <App store={{ ...props.store, saveDefault }} />
  );

  const componentProps = component
    .find(AccountCard)
    .at(1) // bar not default
    .props();

  componentProps.onMakeDefault(componentProps.address);

  expect(saveDefault).toHaveBeenCalled();
});
