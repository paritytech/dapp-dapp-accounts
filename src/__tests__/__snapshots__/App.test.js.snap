// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <Actionbar
    title={
      <FormattedMessage
        defaultMessage="Accounts Visible to Dapps"
        id="dapps.accounts.title"
        values={Object {}}
      />
    }
  />
  <div>
    <Header
      as="h4"
    >
      Default Account
    </Header>
    <span>
      This is your default account to be used in the Dapps. It is also the account shown on the top bar.
    </span>
    <CardGroup>
      <AccountCard
        account={
          Object {
            "meta": Object {},
            "name": "FooName",
          }
        }
        address="foo"
        isDefault={true}
        isVisible={true}
      />
    </CardGroup>
  </div>
  <div>
    <Header
      as="h4"
    >
      Other Accounts
    </Header>
    <span>
      These are the other accounts that can be accessed by Dapps. Toggle them to make them visible/hidden for Dapps.
    </span>
    <CardGroup>
      <AccountCard
        account={
          Object {
            "meta": Object {},
            "name": "BarName",
          }
        }
        address="bar"
        isVisible={true}
        key="bar"
        onMakeDefault={[Function]}
        onToggleVisible={[Function]}
      />
      <AccountCard
        account={
          Object {
            "meta": Object {},
            "name": "BazName",
          }
        }
        address="baz"
        isVisible={false}
        key="baz"
        onMakeDefault={[Function]}
        onToggleVisible={[Function]}
      />
    </CardGroup>
  </div>
</div>
`;

exports[`should render correctly when no account 1`] = `""`;
